test/bidirectional-streams.spec.js

```diff
@@ -1,138 +1,86 @@
 
 /**
  * @template T
  * @typedef {import('../lib/types').Deferred<T>} Deferred<T>
  */
 
-const SERVER_PATH = '/bidirectional-streams'
-
 describe('bidirectional streams', function () {
-  /** @type {import('../lib/server').Http3Server} */
-  let server
-  /** @type {import('./fixtures/certificate.js').Certificate} */
-  let certificate
+  // FIXME: sometimes there are seemingly arbitrary 5s delays in
+  // communicating with the server under node.js
+  this.timeout(30000)
+
   /** @type {import('../lib/dom').WebTransport | undefined} */
   let client
-  /** @type {string} */
-  let url
-
-  this.timeout(30000) // this can vary depending on the setup and is not an error
-  // but it smells like a work around regarding a deeper problem
-
-  beforeEach(async () => {
-    this.timeout(2000)
-    ;({ server, certificate } = await createServer())
-    server.startServer()
-    await server.ready
-
-    const address = server.address()
-
-    if (address == null || address.port == null) {
-      throw new Error('No address')
-    }
-
-    url = `https://${address.host}:${address.port}`
-  })
 
   afterEach(async () => {
     if (client != null) {
       client.close()
-    }
-
-    if (server != null) {
-      server.stopServer()
-      await server.closed
+      client = undefined
     }
   })
 
   it('sends and receives data over an outgoing bidirectional stream', async () => {
-    this.timeout(200)
-    // server context - waits for the client to open a bidi stream and pipes it back to them
-    Promise.resolve().then(async () => {
-      const session = await getReaderValue(server.sessionStream(SERVER_PATH))
-      const bidiStream = await getReaderValue(
-        session.incomingBidirectionalStreams
-      )
-      // redirect input to output
-      await bidiStream.readable.pipeTo(bidiStream.writable)
-    })
-
     // client context - connects to the server, opens a bidi stream, sends some data and reads the response
-    client = new WebTransport(`${url}${SERVER_PATH}`, {
-      serverCertificateHashes: [
-        {
-          algorithm: 'sha-256',
-          value: certificate.hash
-        }
-      ]
-    })
+    client = new WebTransport(
+      `${process.env.SERVER_URL}/bidirectional_client_initiated_echo`,
+      {
+        serverCertificateHashes: [
+          {
+            algorithm: 'sha-256',
+            value: readCertHash(process.env.CERT_HASH)
+          }
+        ]
+      }
+    )
     await client.ready
 
-    const input = [
-      Uint8Array.from([0, 1, 2, 3, 4]),
-      Uint8Array.from([5, 6, 7, 8, 9]),
-      Uint8Array.from([10, 11, 12, 13, 14])
-    ]
-
     const stream = await client.createBidirectionalStream()
-    await writeStream(stream.writable, input)
+    await writeStream(stream.writable, KNOWN_BYTES)
 
-    const output = await readStream(stream.readable)
-    expect(output).to.deep.equal(
-      input,
+    const output = await readStream(stream.readable, KNOWN_BYTES.length)
```

Comment:
KNOWN_BYTES.length is the number of chunks and this causes the test occasionally to fail.
Anyway, and this probaly a bug in the browser, the browser sometimes does not send a fin. So if the browser behaves correctly this should not be an issue, but the browser does not.